src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java

```diff
@@ -0,0 +1,232 @@
+        customObjects.add(createCustomObject("4", "pending", Statistics.RESPONSE_ERROR_PERMANENT));
+        tenantConfiguration.setClient(mockSphereClient(customObjects));
+        Statistics statistic = emailProcessor.processEmails(tenantConfiguration).toCompletableFuture().join();
+        assertEquals(statistic.getProcessedEmails(), 3);
+        assertEquals(statistic.getSuccessfulSendedEmails(), 1);
+        assertEquals(statistic.getTemporarilyErrors(), 1);
+        assertEquals(statistic.getPermanentErrors(), 1);
+
+        customObjects = new ArrayList<CustomObject<JsonNode>>();
+        customObjects.add(createCustomObject("1", "pending", Statistics.RESPONSE_CODE_SUCCESS));
+        customObjects.add(createCustomObject("2", "pending", Statistics.RESPONSE_ERROR_TEMP));
+        customObjects.add(createCustomObject("3", "pending", Statistics.RESPONSE_ERROR_TEMP));
+        customObjects.add(createCustomObject("4", "pending", Statistics.RESPONSE_ERROR_PERMANENT));
+        tenantConfiguration.setClient(mockSphereClient(customObjects));
+        statistic = emailProcessor.processEmails(tenantConfiguration).toCompletableFuture().join();
+        assertEquals(statistic.getProcessedEmails(), 4);
+        assertEquals(statistic.getSuccessfulSendedEmails(), 1);
+        assertEquals(statistic.getTemporarilyErrors(), 2);
+        assertEquals(statistic.getPermanentErrors(), 1);
+
+        customObjects = new ArrayList<CustomObject<JsonNode>>();
+        customObjects.add(createCustomObject("1", "pending", Statistics.RESPONSE_CODE_SUCCESS));
+        customObjects.add(createCustomObject("2", "pending", Statistics.RESPONSE_CODE_SUCCESS));
+        customObjects.add(createCustomObject("3", "pending", Statistics.RESPONSE_CODE_SUCCESS));
+        tenantConfiguration.setClient(mockSphereClient(customObjects));
+        statistic = emailProcessor.processEmails(tenantConfiguration).toCompletableFuture().join();
+        assertEquals(statistic.getProcessedEmails(), 3);
+        assertEquals(statistic.getSuccessfulSendedEmails(), 3);
+        assertEquals(statistic.getTemporarilyErrors(), 0);
+        assertEquals(statistic.getPermanentErrors(), 0);
+    }
+
+
+    @Test
+    public void shouldNotProcessEmails() throws Exception {
+        customObjects.add(createCustomObject("1", "error", Statistics.RESPONSE_CODE_SUCCESS));
+        customObjects.add(createCustomObject("2", "error", Statistics.RESPONSE_ERROR_TEMP));
+        customObjects.add(createCustomObject("3", "error", Statistics.RESPONSE_CODE_SUCCESS));
+        tenantConfiguration.setClient(mockSphereClient(customObjects));
+        Statistics statistic = emailProcessor.processEmails(tenantConfiguration).toCompletableFuture().join();
+        assertEquals(statistic.getProcessedEmails(), 0);
+        assertEquals(statistic.getSuccessfulSendedEmails(), 0);
+        assertEquals(statistic.getTemporarilyErrors(), 0);
+
+    }
+
+    @Test
+    public void shouldNotProcessEmails2() throws Exception {
```

Comment:
Having more self explanatory name for unit tests is much more convenient as the reader/dev won't have to go through the test to find which condition is being tested against. For example, here I can't know under which condition should the emails not be processed, unless I go through the code of the method. However, if the name was `shouldNotProcessEmailsWhenThereAreNotCustomObjects` or with a more popular convention [`MethodName_StateUnderTest_ExpectedBehavior`](https://dzone.com/articles/7-popular-unit-test-naming) -> `processEmails_WithNoCustomObjectsInTenant_ShouldNotProcessEmails` would have saved save time.

Reply:
b7b2ba6